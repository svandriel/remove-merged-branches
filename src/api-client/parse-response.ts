import Cheerio, { Cheerio as CheerioAPI, Element } from 'cheerio';

export interface ParsedResponse {
    count: number;
}

export interface Filter {
    title: string;
    options: FilterOption[];
}

export interface FilterOption {
    title: string;
    value: string;
    subFilters: LeafFilterOption[];
}

export interface LeafFilterOption {
    title: string;
    key: string;
    value: string;
}

export function parseInitialPage(html: string): Filter[] {
    const $ = Cheerio.load(html);
    const results = $('.vacancy-filter-panel')
        .toArray()
        .map(element => {
            const el = $(element);
            const title = el.find('.vacancy-filter-panel__title span').text().trim();
            const options = el
                .find('.vacancy-filter-checkbox__item')
                .toArray()
                .map(item => parseFilterOption(item, $));

            const filter = {
                title,
                options: options.map(option => ({
                    title: option.title,
                    value: option.value,
                    subFilters: option.subFilters,
                })),
            };
            return filter;
        });
    return results;
}

function parseFilterOption(filterOptionItem: Element, $: typeof Cheerio): FilterOption {
    const $filterOptionItem = $(filterOptionItem);
    const result = parseCheckboxItem($filterOptionItem);

    const subFilters = $filterOptionItem
        .find('.vacancy-filter-checkbox-2nd-layer__item')
        .toArray()
        .map(subItem => parseCheckboxItem($(subItem)));

    return { ...result, subFilters };
}

function parseCheckboxItem($item: CheerioAPI<Element>) {
    const input = $item.find('input[type=checkbox]').first();
    const label = $item.find('label').first();
    const id = input.attr('id')?.split('_');
    const title = removeSpaces(label.text());
    const [key, value] = id ?? [];
    return { title, key, value };
}

export function parseIndexPageResponse(html: string): ParsedResponse {
    const $ = Cheerio.load(html);

    const countStr = $('.vacancy-result-bar__totals-badge').text();
    const count = parseInteger(countStr);

    return {
        count,
    };
}

function parseInteger(intString: string): number {
    return +intString.replace(/[.,]/g, '');
}

function removeSpaces(str: string): string {
    return str.replace(/\s+/g, ' ').trim();
}

// function uniq<T>(list: T[]): T[] {
//     return Array.from(new Set(list));
// }
