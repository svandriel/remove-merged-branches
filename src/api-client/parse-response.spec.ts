import fs from 'fs';
import path from 'path';

import { parseInitialPage } from './parse-response';

const testDataFolder = path.join(__dirname, '..', '..', 'test-data');
const contents = fs.readFileSync(path.join(testDataFolder, 'index.html'));

describe('parseResponse', () => {
    describe('parseInitialPage', () => {
        it('works', async () => {
            const result = parseInitialPage(contents.toString());
            expect(result).toMatchSnapshot();
        });
    });
});
