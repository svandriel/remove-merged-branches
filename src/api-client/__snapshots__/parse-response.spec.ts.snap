// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseResponse parseInitialPage works 1`] = `
Array [
  Object {
    "options": Array [
      Object {
        "subFilters": Array [],
        "title": "Administratief/secretarieel",
        "value": "CVG.02",
      },
      Object {
        "subFilters": Array [],
        "title": "Agrarisch",
        "value": "CVG.27",
      },
      Object {
        "subFilters": Array [],
        "title": "Cultureel/taalkundig",
        "value": "CVG.04",
      },
      Object {
        "subFilters": Array [],
        "title": "Documentatie en Informatievoorziening",
        "value": "CVG.32",
      },
      Object {
        "subFilters": Array [],
        "title": "Facilitair/dienstverlening",
        "value": "CVG.09",
      },
      Object {
        "subFilters": Array [],
        "title": "Financieel/economisch",
        "value": "CVG.06",
      },
      Object {
        "subFilters": Array [
          Object {
            "key": "CVG.08-expertises",
            "title": "Architectuur en infrastructuur",
            "value": "Architectuur en infrastructuur",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "Beheer",
            "value": "Beheer",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "(Data)analyse",
            "value": "(Data)analyse",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "Development en testing",
            "value": "Development en testing",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "ICT-(project)management",
            "value": "ICT-(project)management",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "ICT-sourcing",
            "value": "ICT-sourcing",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "Implementatie en service delivery",
            "value": "Implementatie en service delivery",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "Informatiebeleid en strategie",
            "value": "Informatiebeleid en strategie",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "Informatiebeveiliging en audit",
            "value": "Informatiebeveiliging en audit",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "Security",
            "value": "Security",
          },
          Object {
            "key": "CVG.08-expertises",
            "title": "Support",
            "value": "Support",
          },
        ],
        "title": "ICT",
        "value": "CVG.08",
      },
      Object {
        "subFilters": Array [],
        "title": "Inkoop/verkoop",
        "value": "CVG.26",
      },
      Object {
        "subFilters": Array [],
        "title": "Internationaal",
        "value": "CVG.33",
      },
      Object {
        "subFilters": Array [],
        "title": "Juridisch",
        "value": "CVG.10",
      },
      Object {
        "subFilters": Array [],
        "title": "Medisch/verzorging",
        "value": "CVG.12",
      },
      Object {
        "subFilters": Array [],
        "title": "Natuur/milieu",
        "value": "CVG.28",
      },
      Object {
        "subFilters": Array [],
        "title": "Onderwijs/opleiding",
        "value": "CVG.15",
      },
      Object {
        "subFilters": Array [],
        "title": "Onderzoek/wetenschap",
        "value": "CVG.14",
      },
      Object {
        "subFilters": Array [],
        "title": "Opslag/vervoer/logistiek",
        "value": "CVG.16",
      },
      Object {
        "subFilters": Array [],
        "title": "Orde/vrede/veiligheid",
        "value": "CVG.31",
      },
      Object {
        "subFilters": Array [],
        "title": "Personeel en organisatie",
        "value": "CVG.18",
      },
      Object {
        "subFilters": Array [],
        "title": "Reclame/communicatie/marketing/pr/voorlichting",
        "value": "CVG.20",
      },
      Object {
        "subFilters": Array [],
        "title": "Sociaal/maatschappelijk/welzijn",
        "value": "CVG.22",
      },
      Object {
        "subFilters": Array [],
        "title": "Stedenbouwkundig/ruimtelijke ordening",
        "value": "CVG.29",
      },
      Object {
        "subFilters": Array [],
        "title": "Techniek/productie",
        "value": "CVG.24",
      },
      Object {
        "subFilters": Array [],
        "title": "Telecom",
        "value": "CVG.07",
      },
      Object {
        "subFilters": Array [],
        "title": "Verkeer/nautisch",
        "value": "CVG.30",
      },
    ],
    "title": "Vakgebied",
  },
  Object {
    "options": Array [
      Object {
        "subFilters": Array [],
        "title": "Gepromoveerd",
        "value": "CWD.02",
      },
      Object {
        "subFilters": Array [],
        "title": "Master/doctoraal",
        "value": "CWD.04",
      },
      Object {
        "subFilters": Array [],
        "title": "Bachelor - HBO",
        "value": "CWD.08",
      },
      Object {
        "subFilters": Array [],
        "title": "MBO",
        "value": "CWD.12",
      },
      Object {
        "subFilters": Array [],
        "title": "VWO",
        "value": "CWD.13",
      },
      Object {
        "subFilters": Array [],
        "title": "HAVO",
        "value": "CWD.15",
      },
      Object {
        "subFilters": Array [],
        "title": "VMBO / MAVO / LBO",
        "value": "CWD.16",
      },
    ],
    "title": "Werk-/Denkniveau",
  },
  Object {
    "options": Array [],
    "title": "Afstand",
  },
  Object {
    "options": Array [
      Object {
        "subFilters": Array [],
        "title": "Beleid / Staf/ Advies / Onderzoek / Consultancy",
        "value": "CFT.04",
      },
      Object {
        "subFilters": Array [],
        "title": "Bestuurlijk",
        "value": "CFT.06",
      },
      Object {
        "subFilters": Array [],
        "title": "Inspectie / Toezicht",
        "value": "CFT.10",
      },
      Object {
        "subFilters": Array [],
        "title": "Management / Leidinggevend",
        "value": "CFT.02",
      },
      Object {
        "subFilters": Array [],
        "title": "Onderwijsgevend",
        "value": "CFT.09",
      },
      Object {
        "subFilters": Array [],
        "title": "Uitvoerend / Ondersteunend",
        "value": "CFT.08",
      },
    ],
    "title": "Functietype",
  },
  Object {
    "options": Array [
      Object {
        "subFilters": Array [],
        "title": "Vast dienstverband",
        "value": "CSD.02",
      },
      Object {
        "subFilters": Array [],
        "title": "Tijdelijk dienstverband",
        "value": "CSD.04",
      },
      Object {
        "subFilters": Array [],
        "title": "Tijdelijk dienstverband met uitzicht op vast",
        "value": "CSD.10",
      },
      Object {
        "subFilters": Array [],
        "title": "Trainee",
        "value": "CSD.06",
      },
      Object {
        "subFilters": Array [],
        "title": "Oproepkracht",
        "value": "CSD.14",
      },
      Object {
        "subFilters": Array [],
        "title": "Detachering",
        "value": "CSD.12",
      },
    ],
    "title": "Dienstverband",
  },
  Object {
    "options": Array [
      Object {
        "subFilters": Array [],
        "title": "Algemene Inlichtingen- en Veiligheidsdienst",
        "value": "01840",
      },
      Object {
        "subFilters": Array [],
        "title": "Algemene Rekenkamer",
        "value": "01020",
      },
      Object {
        "subFilters": Array [],
        "title": "Autoriteit Persoonsgegevens",
        "value": "01140",
      },
      Object {
        "subFilters": Array [],
        "title": "Belastingdienst",
        "value": "01130",
      },
      Object {
        "subFilters": Array [],
        "title": "Centraal Bureau voor de Statistiek",
        "value": "01090",
      },
      Object {
        "subFilters": Array [],
        "title": "Dienst Justitiële Inrichtingen",
        "value": "01481",
      },
      Object {
        "subFilters": Array [],
        "title": "Dopingautoriteit",
        "value": "50099",
      },
      Object {
        "subFilters": Array [],
        "title": "Douane",
        "value": "50209",
      },
      Object {
        "subFilters": Array [],
        "title": "Eerste Kamer der Staten-Generaal",
        "value": "01410",
      },
      Object {
        "subFilters": Array [],
        "title": "Kabinet van de Koning",
        "value": "01980",
      },
      Object {
        "subFilters": Array [],
        "title": "Kamer van Koophandel",
        "value": "01830",
      },
      Object {
        "subFilters": Array [],
        "title": "Kanselarij der Nederlandse Orden",
        "value": "01930",
      },
      Object {
        "subFilters": Array [],
        "title": "Kansspelautoriteit",
        "value": "30080",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Algemene Zaken",
        "value": "01030",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Binnenlandse Zaken en Koninkrijksrelaties",
        "value": "01060",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Buitenlandse Zaken",
        "value": "01070",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Defensie",
        "value": "01310",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Economische Zaken en Klimaat",
        "value": "01420",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Financiën",
        "value": "01430",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Infrastructuur en Waterstaat",
        "value": "01920",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Justitie en Veiligheid",
        "value": "01480",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Landbouw, Natuur en Voedselkwaliteit",
        "value": "50079",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Onderwijs, Cultuur en Wetenschap",
        "value": "01630",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Sociale Zaken en Werkgelegenheid",
        "value": "01820",
      },
      Object {
        "subFilters": Array [],
        "title": "Ministerie van Volksgezondheid, Welzijn en Sport",
        "value": "01940",
      },
      Object {
        "subFilters": Array [],
        "title": "Nationale Ombudsman",
        "value": "01580",
      },
      Object {
        "subFilters": Array [],
        "title": "Onderzoeksraad voor Veiligheid",
        "value": "01860",
      },
      Object {
        "subFilters": Array [],
        "title": "Raad van State",
        "value": "01680",
      },
      Object {
        "subFilters": Array [],
        "title": "Rechtspraak",
        "value": "01700",
      },
      Object {
        "subFilters": Array [],
        "title": "Rijkstraineeprogramma",
        "value": "01770",
      },
      Object {
        "subFilters": Array [],
        "title": "Rijkswaterstaat",
        "value": "01380",
      },
      Object {
        "subFilters": Array [],
        "title": "Toeslagen",
        "value": "50213",
      },
      Object {
        "subFilters": Array [],
        "title": "Tweede Kamer der Staten-Generaal",
        "value": "01900",
      },
      Object {
        "subFilters": Array [],
        "title": "Zorginstituut Nederland",
        "value": "11002",
      },
    ],
    "title": "Werkgever",
  },
  Object {
    "options": Array [
      Object {
        "subFilters": Array [],
        "title": "Schaal 1",
        "value": "1",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 2",
        "value": "2",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 3",
        "value": "3",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 4",
        "value": "4",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 5",
        "value": "5",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 6",
        "value": "6",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 7",
        "value": "7",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 8",
        "value": "8",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 9",
        "value": "9",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 10",
        "value": "10",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 11",
        "value": "11",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 12",
        "value": "12",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 13",
        "value": "13",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 14",
        "value": "14",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 15",
        "value": "15",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 16",
        "value": "16",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 17",
        "value": "17",
      },
      Object {
        "subFilters": Array [],
        "title": "Schaal 18",
        "value": "18",
      },
      Object {
        "subFilters": Array [],
        "title": "Niveau 19",
        "value": "19",
      },
    ],
    "title": "Salarisniveau",
  },
  Object {
    "options": Array [],
    "title": "Gepubliceerd sinds",
  },
]
`;
